File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/core/routing/RoutingHandlerTest.java
package uk.nhs.nhsx.core.routing;

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
import org.junit.Test;
import uk.nhs.nhsx.ContextBuilder;
import uk.nhs.nhsx.core.HttpResponses;

import java.util.HashMap;
import java.util.Optional;

import static com.google.common.collect.Maps.newHashMap;

public class RoutingHandlerTest {

    @Test(expected = RuntimeException.class)
    public void noHeaders() throws Exception {
        new MyRoutingHandler("content-type").handleRequest(new APIGatewayProxyRequestEvent(), new ContextBuilder.TestContext());
    }

    @Test
    public void lowercase() throws Exception {
        APIGatewayProxyRequestEvent request = new APIGatewayProxyRequestEvent();
        HashMap<String, String> headers = newHashMap();
        headers.put("Content-Type", "something");
        request.setHeaders(headers);
        new MyRoutingHandler("content-type").handleRequest(request, new ContextBuilder.TestContext());
    }

    @Test
    public void uppercase() throws Exception {
        APIGatewayProxyRequestEvent request = new APIGatewayProxyRequestEvent();
        HashMap<String, String> headers = newHashMap();
        headers.put("content-type", "something");
        request.setHeaders(headers);
        new MyRoutingHandler("Content-Type").handleRequest(request, new ContextBuilder.TestContext());
    }

    private static class MyRoutingHandler extends RoutingHandler {

        public final String header;

        private MyRoutingHandler(String header) {
            this.header = header;
        }

        @Override
        public Routing.Handler handler() {
            return request -> Optional.ofNullable(request.getHeaders().get(header))
                .map(h -> HttpResponses.ok())
                .orElseThrow(() -> new RuntimeException("not there"));
        }
    }
}

File: src/aws/lambdas/incremental_distribution/src/test/java/uk/nhs/nhsx/core/routing/RoutingHandlerTest.kt
package uk.nhs.nhsx.core.routing

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import uk.nhs.nhsx.core.HttpResponses
import uk.nhs.nhsx.testhelper.ContextBuilder.TestContext
import java.util.Optional

class RoutingHandlerTest {

    @Test
    fun noHeaders() {
        assertThrows(RuntimeException::class.java) {
            MyRoutingHandler("content-type").handleRequest(APIGatewayProxyRequestEvent(), TestContext())
        }
    }

    @Test
    fun lowercase() {
        val request = APIGatewayProxyRequestEvent()
        val headers = mutableMapOf<String, String>()
        headers["Content-Type"] = "something"
        request.headers = headers
        MyRoutingHandler("content-type").handleRequest(request, TestContext())
    }

    @Test
    fun uppercase() {
        val request = APIGatewayProxyRequestEvent()
        val headers = mutableMapOf<String, String>()
        headers["content-type"] = "something"
        request.headers = headers
        MyRoutingHandler("Content-Type").handleRequest(request, TestContext())
    }

    private class MyRoutingHandler(val header: String) : RoutingHandler() {
        override fun handler(): ApiGatewayHandler =
            ApiGatewayHandler { request, _ ->
                Optional.ofNullable(request.headers[header])
                    .map { HttpResponses.ok() }
                    .orElseThrow { RuntimeException("not there") }
            }
    }
}
